src/app/Library/CrudPanel/Traits/Fields.php

```diff
@@ -56,6 +56,18 @@ public function makeSureFieldHasNecessaryAttributes($field)
 
         $this->setupFieldValidation($field, $field['parentFieldName'] ?? false);
 
+        if (isset($field['relation_type']) && $field['relation_type'] === 'MorphTo') {
+            [$morphTypeFieldName, $morphIdFieldName] = $this->getMorphToFieldNames($field['name']);
+            if (! $this->hasFieldWhere('name', $morphTypeFieldName) || ! $this->hasFieldWhere('name', $morphIdFieldName)) {
+                // create the morph fields in the crud panel
+                $field = $this->createMorphToRelationFields($field, $morphTypeFieldName, $morphIdFieldName);
+                foreach ($field['morphOptions'] ?? [] as $morphOption) {
+                    [$key, $label, $options] = $this->getMorphOptionStructured($morphOption);
+                    $field = $this->addMorphOption($field, $key, $label, $options);
+                }
+            }
+        }
```

Comment:
Ugh... this kind of pollutes this function, doesn't it? Shouldn't we move all this logic to a private function, something like `makeSureMorphSubfieldsAreDefined($field)`?

Reply:
Additionally, why do we do this AFTER we call `setupFieldValidation()` and not BEFORE it? Any reason?

Reply:
No special reason that I can recall, it was just so that this block of code was not messing with the clean defined stuff. 

Moving this to a private method solved that issue. 

Cheers